### ISSUE TITLE
RVM - Error running '__rvm_make -j8' while running rvm install on Debian 12

### ISSUE THREAD

**Issue by aarcos** [rvm/rvm#5426]
## Description
First sorry by my english

When try run "rvm install ruby" my system debian 12 with xfce show this error:

~$ rvm install ruby
ruby-3.0.0 - #removing src/ruby-3.0.0..
Checking requirements for debian.
Requirements installation successful.
Installing Ruby from source to: /home/arcosaf/.rvm/rubies/ruby-3.0.0, this may take a while depending on your cpu(s)...
ruby-3.0.0 - #downloading ruby-3.0.0, this may take a while depending on your connection...
ruby-3.0.0 - #extracting ruby-3.0.0 to /home/arcosaf/.rvm/src/ruby-3.0.0.....
ruby-3.0.0 - #configuring.........................................................................
ruby-3.0.0 - #post-configuration..
ruby-3.0.0 - #compiling......................................................................................................................................................................-
Error running '__rvm_make -j8',
please read /home/"user"/.rvm/log/1703800072_ruby-3.0.0/make.log

There has been an error while running make. Halting the installation.
## Steps to reproduce

I use this page for installation ruby with rvm on debian

https://linux.how2shout.com/how-to-install-ruby-version-manager-rvm-on-debian-11-bullseye/

Below here paste portion make.log:

compiling openssl_missing.c
compiling ossl.c
compiling objspace.c
In file included from ossl.h:156,
from ossl.c:10:
openssl_missing.h:239: warning: "TS_VERIFY_CTS_set_certs" redefined
239 | # define TS_VERIFY_CTS_set_certs(ctx, crts) ((ctx)->certs=(crts))
|
In file included from ossl.h:31:
/usr/include/openssl/ts.h:424: note: this is the location of the previous definition
424 | # define TS_VERIFY_CTS_set_certs(ctx, cert) TS_VERIFY_CTX_set_certs(ctx,cert)
|
ossl.c: In function ‘ossl_clear_error’:
ossl.c:311:9: warning: ‘ERR_get_error_line_data’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
311 | while ((e = ERR_get_error_line_data(&file, &line, &data, &flags))) {
| ^~~~~

But after an analysis:

~$ rvm install ruby --with-openssl-dir=/usr/bin/openssl
ruby-3.0.0 - #removing src/ruby-3.0.0..
Checking requirements for debian.
Requirements installation successful.
Installing Ruby from source to: /home/"user"/.rvm/rubies/ruby-3.0.0, this may take a while depending on your cpu(s)...
ruby-3.0.0 - #downloading ruby-3.0.0, this may take a while depending on your connection...
ruby-3.0.0 - #extracting ruby-3.0.0 to /home/"user"/.rvm/src/ruby-3.0.0.....
ruby-3.0.0 - #configuring.........................................................................
ruby-3.0.0 - #post-configuration..
ruby-3.0.0 - #compiling......................................................................................................................................................................|
ruby-3.0.0 - #installing.....................
ruby-3.0.0 - #making binaries executable...
Installed rubygems 3.2.3 is newer than 3.0.9 provided with installed ruby, skipping installation, use --force to force installation.
ruby-3.0.0 - #gemset created /home/"user"/.rvm/gems/ruby-3.0.0@global
ruby-3.0.0 - #importing gemset /home/"user"/.rvm/gemsets/global.gems..............................there was an error installing gem rubygems-bundler
.............................
ruby-3.0.0 - #generating global wrappers........
ruby-3.0.0 - #gemset created /home/"user"/.rvm/gems/ruby-3.0.0
ruby-3.0.0 - #importing gemsetfile /home/"user"/.rvm/gemsets/default.gems evaluated to empty gem list
ruby-3.0.0 - #generating default wrappers........
ruby-3.0.0 - #adjusting #shebangs for (gem irb erb ri rdoc testrb rake).
Install of ruby-3.0.0 - #complete
Ruby was built without documentation, to build it run: rvm docs generate-ri

I hope it helps you to correct the problem
I'm sorry if this is not the right place to put this note, in my favor it is the only site that has a solution for macOS

Best

**Comment by arthurduarte1** [comment]
Same here for ruby 2.7.0 on Mac OS X.

This code below saved my day:
```
brew install openssl@1.1
brew uninstall --ignore-dependencies openssl@3
rvm install 2.7.0
brew install --ignore-dependencies openssl@3
```

The "problem" was about openssl.
Maybe you can look for something similar on linux.

**Comment by aarcos** [comment]
Hi Arthur

Yess the problem is in rvm program.

When run rvm install Ruby does not find openssl directory installation,
this generate the error.

When add to command

rvm install ruby --openssl-dir=

The problem was solved.

Thanks Arthur Best new year

Alejandro

**Comment by arthurduarte1** [comment]
Good to know.
Take care, best for you too. :)

**Comment by brodjustice** [comment]
Trying to get any version below ruby 3 to work with rvm on Debian12 / Ubuntu 22 proved very painful for me. The only method that worked for was to install openssl 1.1.1 alongside openssl 3. For the record this is what I did:

```
wget https://www.openssl.org/source/old/1.1.1/openssl-1.1.1v.tar.gz
tar xf openssl-1.1.1v.tar.gz
cd openssl-1.1.1v
./config --prefix=/usr/local/openssl1.1.1v --openssldir=/usr/local/openssl1.1.1v/etc
make
sudo make install
```

Now rebuild the rvm ruby, eg:

`
rvm reinstall 2.7.8 --with-openssl-dir=/usr/local/openssl1.1.1v`

**Comment by monfresh** [comment]
@brodjustice That is expected. Ruby versions older than 3.1 do not work with OpenSSL 3. They require OpenSSL 1.1. This is a common point of confusion, so I wrote this blog post with a handy compatibility chart:
https://www.rubyonmac.dev/openssl-versions-supported-by-ruby

**Comment by brodjustice** [comment]
@monfresh  I think the point is that Linux users were have been used to installing RVM on top of their fresh Linux installation and within minutes are good to go for any Ruby between 1.9 and 3.x. This is no longer the case, RVM is effectively broken. There are many issues that seem to have the missing openssl version as the core problem https://github.com/rvm/rvm/issues/5434, https://github.com/rvm/rvm/issues/5361#issuecomment-1612087695,  https://github.com/rvm/rvm/issues/5074, etc.

**Comment by monfresh** [comment]
Ah, yes. I agree. The good news is that there are alternatives available, such as asdf, rbenv, and chruby/ruby-install. Based on my experience on macOS, they are more reliable than rvm.

**Comment by ahummel25** [comment]
> @monfresh I think the point is that Linux users were have been used to installing RVM on top of their fresh Linux installation and within minutes are good to go for any Ruby between 1.9 and 3.x. This is no longer the case, RVM is effectively broken. There are many issues that seem to have the missing openssl version as the core problem #5434, [#5361 (comment)](https://github.com/rvm/rvm/issues/5361#issuecomment-1612087695), #5074, etc.

This is what I'm still experiencing on mac. Does this effectively mean there isn't a workaround? I can no longer get any `3.x` version to install even when providing the `--with-openssl-dir` flag.

I've tried numerous install commands including the below (with all sorts of flags provided, but just listing these two for brevity's sake):
```zsh
rvm install 3.3.2 --with-openssl-dir=$(brew --prefix openssl@3)
```
```zsh
rvm install 3.3.2 --with-openssl-dir=$(brew --prefix openssl@3.3)
```

The logs seem to confirm that it's picking up on openssl 3.x but it always fails with something like `error running '__rvm_make -j16'`.

Is RVM essentially broken until further notice?

**Comment by brodjustice** [comment]
I'm not familiar with Mac set-up for RVM. But I can say that [my work around here](https://github.com/rvm/rvm/issues/5426#issuecomment-1902266154) still works on Linux.

I'm not sure I'd say that "RVM essentially broken", on Linux it depends on the Ruby and OS/SSL combination you want use as @monfresh indicates [here](https://github.com/rvm/rvm/issues/5426#issuecomment-1905265052)

**Comment by MattCrl** [comment]
@brodjustice Thanks to your solution I could install an old Ruby version on Debian bookworm, but now when running `bundle install`, I got this certificate error :
```
Could not verify the SSL certificate for https://rubygems.org/quick/Marshal.4.8/rake-10.5.0.gemspec.rz.
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. 
For information about OpenSSL certificates, see http://bit.ly/ruby-ssl. 
To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.
```
Have you encountered this issue too ?

**Comment by brodjustice** [comment]
@MattCrl Yes I believe I had a similar issue and it was related to incompatible versions of OpenSSL. I'm sorry, I can't recall how I fixed it in my case but maybe this helps: https://github.com/rubygems/bundler/issues/3107

**Comment by raelgc** [comment]
This is not a RVM issue. Rubies lower than 3.1 requires OpenSSL 1.1, which is not available in modern OSs. The solutions is like pointed above: download OpenSSL 1.1., build it locally, **point to correct the cert files** (missing above), and build Ruby again.

It should be like this (tested on Ubuntu 22 and 24):

```
cd ~/Downloads
wget https://www.openssl.org/source/openssl-1.1.1w.tar.gz
tar zxvf openssl-1.1.1w.tar.gz
cd openssl-1.1.1w
./config --prefix=$HOME/.openssl/openssl-1.1.1w --openssldir=$HOME/.openssl/openssl-1.1.1w
make
make install
rm -rf ~/.openssl/openssl-1.1.1w/certs
ln -s /etc/ssl/certs ~/.openssl/openssl-1.1.1w/certs
cd ~
```

Now we can proceed and install Ruby (example with Ruby 3.0.4):


    rvm install ruby-3.0.4 --with-openssl-dir=$HOME/.openssl/openssl-1.1.1w

**Comment by krulun** [comment]
> Same here for ruby 2.7.0 on Mac OS X.
> 
> This code below saved my day:
> 
> ```
> brew install openssl@1.1
> brew uninstall --ignore-dependencies openssl@3
> rvm install 2.7.0
> brew install --ignore-dependencies openssl@3
> ```
> 
> The "problem" was about openssl. Maybe you can look for something similar on linux.

nice answer

**Comment by brodjustice** [comment]
@MattCrl 

> Have you encountered this issue too ?

Yes, I just now I needed to run an old Ruby 2.7 app. This required a specific bundler version:
```
gem install bundler -v '2.3.26'
bundle _2.3.26_ install
```
